services/twilio_verify.js

```diff
@@ -0,0 +1,51 @@
+const twilio = require('twilio');
+const password = require('secure-random-password');
+
+const ACCOUNT_SID = process.env.TWILIO_ACCOUNT_SID;
+const AUTH_TOKEN = process.env.TWILIO_AUTH_TOKEN;
+const VERIFY_SERVICE_SID = process.env.TWILIO_VERIFY_SERVICE_SID;
+const USE_MOCK_TWILIO = process.env.USE_MOCK_TWILIO === 'true';
+
+const VERIFY_DELIVERY_CHANNEL = 'sms';
+
+module.exports = class TwilioVerifyService {
```

Comment:
Do you think there would be any value in having separate classes for mock vs live instances? This is kind of an unusual case in that the interface is simple but every method needs to check whether it should be mocking and effectively does something completely different.

Curious to hear your thoughts as to whether the overhead of a "duplicate" class implementing the same interface would be a good tradeoff here. This is something that might be more clear in TypeScript with actual `interface`s.

Reply:
I thought about this for a while too and do think a separate class is the ideal case but we don't have a good way of injecting or overriding the live instance when we'd like. Two approaches I thought of were:

1. Using Koa context:
```js
// in app.js

app.context.services = {
  twilioVerify: new TwilioVerifyService(),
} ;

// later somewhere, also need to do it for `test`
if (process.env.NODE_ENV === 'development') {
  app.context.services = {
    twilioVerify: new TwilioVerifyMockService(),
  } ;
}
```

2. Conditional `require`
```js
// in app.js

const TwilioVerifyService = process.env.NODE_ENV === 'development' ? require('mock') : require('real');
```

2 wouldn't work for `test` since it's hardcoded in module scope but 1. Could. It then became more confusing because the live and mock classes need different arguments:
```js
// live
send({ to }) {
    return this.verifyService.verifications.create({ to, channel: VERIFY_DELIVERY_CHANNEL });
}

// mock
send({ to }, emitServerEvent) {
    const securityCode = this.generateMockSecurityCode();

    this.securityCodes[to] = securityCode;
    emitServerEvent(securityCode);
}
```
So we'd always need to pass in `emitServerEvent` but we'd actually only be used _if_ we override the live instance with the mock.

This is sort of why I was more inclined to merge the two together but i'm still on the fence, keen to hear your thoughts on the above!

Reply:
Hmm good points, I like the `context` override solution but it may not be very durable if we move off of Koa later. Another non-invasive way might be something like:
```
module.exports = {
    TwilioService: useMockService ? MockService : TwilioService,
};
```
but that introduces a small inconsistency with the services that use a default export. The `emitServerEvent` is trickier but maybe could be worked into the mock's constructor... still not great though I think.

In this case it might be easier to just rely on the simplicity of the interface, it's not like it's particularly hard to reason about as is 😄

Reply:
I think we can do:
```js
module.exports = USE_TWILIO_MOCK ? TwilioServiceMock : TwilioService;
```
Still have the issue of arguments though 😅

Reply:
Starting to think that's a cleaner option ⬆️  should create better separation and with both classes in the same file the arguments issue shouldn't be too complicated to figure out 😄  i'll push up a change.